### PR TITLE
terraform: share graph walker's variables lock w/ interpolater

### PR DESCRIPTION
**Pull request by phinze**
The ContextGraphWalker struct includes a lock that's passed down to
BuiltinEvalContext and guards access to interpolation variables as
they're written using SetVariables.

The likely problem being expressed in #5733 is that the same map
reference is also passed down to the Interpolater.Variables field, which
is used for variable lookup.

Here, we plumb the same lock we're using to guard access for writes down
and acquire it before doing variable reads as well. It's not as fine
grained as perhaps it could be, but all the context tests pass and I
believe this should address #5733.

### CODE REVIEW

File: terraform/interpolate.go [hunk 1]
```diff
@@ -273,6 +274,8 @@ func (i *Interpolater) valueUserVar(
 	n string,
 	v *config.UserVariable,
 	result map[string]ast.Variable) error {
+	i.VariablesLock.Lock()
+	defer i.VariablesLock.Unlock()
```
Comment (mitchellh) [inline terraform/interpolate.go]: You probably don't need to defer here, just unlock right below the next line.

**Comment by mitchellh** [comment]
One comment, then go for merge.

**Comment by phinze** [comment]
There's a `range` over `i.Variables` on line 298 I thought I needed to cover as well?

**Comment by mitchellh** [comment]
Oh I missed it. Sounds good.

**Pull request merged**